### PR TITLE
[SPARK-18069][PYTHON] Make PySpark doctests for SQL self-contained

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to make the examples in Python API documentation as self-contained.

These seem pretty common in Python API documentation, for example, [pandas](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pivot.html#pandas.DataFrame.pivot) and [numpy](https://docs.scipy.org/doc/numpy/reference/arrays.ndarray.html).

**Before**

**After**

## How was this patch tested?

Manually tested the doctests.

```
 ./run-tests.py --python-executables=python2.6 --modules=pyspark-sql
 ./run-tests.py --python-executables=python3.6 --modules=pyspark-sql
```

Closes #15053

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @holden, @srowen and @mortada who were in the PR. Could three of you take a look when you have some time?

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -1191,6 +1287,12 @@ def dropna(self, how='any', thresh=None, subset=None):
             This overwrites the `how` parameter.
         :param subset: optional list of column names to consider.
 
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Ops, let me remove this extra newline when I happen to push more commits.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -218,8 +228,10 @@ def writeStream(self):
     def schema(self):
         """Returns the schema of this :class:`DataFrame` as a :class:`pyspark.sql.types.StructType`.
 
+        >>> from pyspark.sql import Row
+        >>> df = spark.createDataFrame([Row(name='Alice', age=2), Row(name='Bob', age=5)])
         >>> df.schema
-        StructType(List(StructField(age,IntegerType,true),StructField(name,StringType,true)))
+        StructType(List(StructField(age,LongType,true),StructField(name,StringType,true)))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: This happened to be `LongType` as it seems numbers become a long type by default via PySpark.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -233,11 +245,13 @@ def schema(self):
     def printSchema(self):
         """Prints out the schema in the tree format.
 
+        >>> from pyspark.sql import Row
+        >>> df = spark.createDataFrame([Row(name='Alice', age=2), Row(name='Bob', age=5)])
         >>> df.printSchema()
         root
-         |-- age: integer (nullable = true)
+         |-- age: long (nullable = true)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: This one too.

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -247,9 +261,11 @@ def explain(self, extended=False):
 
         :param extended: boolean, default ``False``. If ``False``, prints only the physical plan.
 
+        >>> from pyspark.sql import Row
+        >>> df = spark.createDataFrame([Row(name='Alice', age=2), Row(name='Bob', age=5)])
         >>> df.explain()
         == Physical Plan ==
-        Scan ExistingRDD[age#0,name#1]
+        Scan ExistingRDD[age#...,name#...]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: We now create multiple dataframes so the columns are not always `#0` and `#1`.

File: python/pyspark/sql/dataframe.py [hunk 5]
```diff
@@ -296,8 +312,10 @@ def show(self, n=20, truncate=True):
             If set to a number greater than one, truncates long strings to length ``truncate``
             and align cells right.
 
+        >>> from pyspark.sql import Row
+        >>> df = spark.createDataFrame([Row(name='Alice', age=2), Row(name='Bob', age=5)])
         >>> df
-        DataFrame[age: int, name: string]
+        DataFrame[age: bigint, name: string]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: This is the same. It seems `bigint` by default if we don't explicitly provide the schema.

File: python/pyspark/sql/dataframe.py [hunk 6]
```diff
@@ -553,10 +594,10 @@ def repartition(self, numPartitions, *cols):
         +---+-----+
         |age| name|
         +---+-----+
-        |  2|Alice|
         |  5|  Bob|
-        |  2|Alice|
         |  5|  Bob|
+        |  2|Alice|
+        |  2|Alice|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: It seems the order is varied dependent on how we create the dataframe (as I saw the related JIRA I can point out if anyone wants). I haven't looked into this deeper as this variant seems not related to the tests in `repartition` API.

File: python/pyspark/sql/session.py [hunk 7]
```diff
@@ -51,8 +51,10 @@ def toDF(self, schema=None, sampleRatio=None):
         :param samplingRatio: the sample ratio of rows used for inferring
         :return: a DataFrame
 
+        >>> from pyspark.sql import Row
+        >>> rdd = sc.parallelize([Row(field1=1, field2="row1"), Row(field1=2, field2="row2")])
         >>> rdd.toDF().collect()
-        [Row(name=u'Alice', age=1)]
+        [Row(field1=1, field2=u'row1'), Row(field1=2, field2=u'row2')]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: It seems doctest does not run this as the results were already wrong. 

```python
>>> rdd = sc.parallelize(
...          [Row(field1=1, field2="row1"),
...           Row(field1=2, field2="row2"),
...           Row(field1=3, field2="row3")])
>>> rdd.collect()
[Row(field1=1, field2='row1'), Row(field1=2, field2='row2'), Row(field1=3, field2='row3')]
```

**Comment by HyukjinKwon** [comment]
gentle ping @holdenk (somehow writing `@holdenk` does not show your name ... )

**Comment by holdenk** [comment]
Just back from Spark Summit East, I'll try and take a look soon :)

**Comment by HyukjinKwon** [comment]
@davies and @holdenk, what do you think about this change? I am fine if we can't decide it is worth for now. I can close this if any of you is not confident of this change.

**Comment by holdenk** [comment]
I'm slightly against the work to make this change happen and would rather focus on some other Python PRs - I'm not sure it improves readability of the test cases as much as planned but I'll defer to @davies judgement on this one.

**Comment by HyukjinKwon** [comment]
Thanks @holdenk, let me close this for now.

@davies, please give me your opinion. If you think it is worth I will reopen. If not, I will resolve the JIRA.

**Pull request closed without merging**